packages/@aws-cdk/aws-iam/lib/role.ts

```diff
@@ -175,20 +189,66 @@ export class Role extends Resource implements IRole {
           grantee,
           actions,
           resourceArns: [this.roleArn],
-          scope: this
+          scope: this,
         });
       }
+    }
 
-      /**
-       * Grant permissions to the given principal to pass this role.
-       */
-      public grantPassRole(identity: IPrincipal): Grant {
-        return this.grant(identity, 'iam:PassRole');
+    const roleAccount = parsedArn.account;
+
+    class MutableImport extends Import {
+      private readonly attachedPolicies = new AttachedPolicies();
+      private defaultPolicy?: Policy;
+
+      public addToPolicy(statement: PolicyStatement): boolean {
+        if (!this.defaultPolicy) {
+          this.defaultPolicy = new Policy(this, 'Policy');
+          this.attachInlinePolicy(this.defaultPolicy);
+        }
+        this.defaultPolicy.addStatements(statement);
+        return true;
+      }
+
+      public attachInlinePolicy(policy: Policy): void {
+        const policyStack = Stack.of(policy);
+        const policyAccount = policyStack.account;
+
+        // this is complicated, so I extracted the condition to a separate variable
+        // we should attach the policy if...
+        const shouldAttachPolicy =
+          // ...the policy is in the same stack as the role was imported into...
+          policyStack === scopeStack ||
+          // ...or the policy is from an env-agnostic stack...
+          Token.isUnresolved(policyAccount) ||
+          // ...or the ARN with which the role was imported with is dynamic...
+          Token.isUnresolved(roleAccount) ||
+          // ...or the policy is in the same account as the imported role
+          policyAccount === roleAccount;
+
+        if (shouldAttachPolicy) {
+          this.attachedPolicies.attach(policy);
+          policy.attachToRole(this);
+        }
+      }
+    }
+
+    class ImmutableImport extends Import {
+      public addToPolicy(_statement: PolicyStatement): boolean {
+        return false;
+      }
+
+      public attachInlinePolicy(_policy: Policy): void {
+        // do nothing
       }
     }
 
-    return new Import(scope, id);
+    const stackAccount = scopeStack.account;
 
+    return options.mutable !== false &&
+        (Token.isUnresolved(stackAccount) || Token.isUnresolved(roleAccount) ||
```

Comment:
Feels like this is mostly the same check as is happening inside `attachInlinePolicy` (though we're looking at different scopes, obviously). But shouldn't we factor out both complex conditionals into a helper function?

Reply:
Yep. Extracted to a function.